### PR TITLE
[FEATURE] Schedule Task

### PR DESCRIPTION
**Pull request by nickiaconis**
Resolves #14.

- [x] scheduleTask implementation
- [x] Test cases
- [x] ~How to handle Ember < 2.8~ (not needed anymore)
- [x] ~Should we optimize during build for Ember >= 2.8~ (not needed anymore)
- [x] ~Incorporate @mixonic's feedback from #14~
- [x] ~Incorporate nickiaconis/ember-lifeline#1~
- [x] Documentation
- [ ] ¯\\\_(ツ)_/¯

### CODE REVIEW

**Comment by nickiaconis** [comment]
I can't seem to get ember-try working locally. I'm going to add some commits to the PR to test a few things, but I don't think they should remain part of the PR. I'll update once I've removed them.

**Comment by nickiaconis** [comment]
This is weird. I added a scenario for Ember LTS 2.4 in order to verify the [run.schedule patch](https://github.com/nickiaconis/ember-lifeline/commit/f0c3d91b15add447814d52e540ef04aa823f803a) I'd included. However, the new scenario is [passing](https://travis-ci.org/rwjblue/ember-lifeline/jobs/276892079) without the patch.

Does anyone have insight as to why?

**Comment by scalvert** [comment]
How is ember-try failing for you locally? The only issues I've had with it recently is in conjunction with the [2.12 + ember-source scenario](https://travis-ci.org/scalvert/ember-sinon-sandbox/builds/273711274).

**Comment by nickiaconis** [comment]
@scalvert It's breaking like this:

```
ember try:one ember-lts-2.4
yarn install v1.0.2
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
warning "ajv-keywords@1.5.1" has incorrect peer dependency "ajv@>=4.10.0".
[4/4] 📃  Building fresh packages...
✨  Done in 142.64s.
npm WARN gentlyRm not removing /Users/niaconis/github-workspace/ember-lifeline/node_modules/.bin/rimraf as it wasn't installed by /Users/niaconis/github-workspace/ember-lifeline/node_modules/rimraf
npm WARN gentlyRm not removing /Users/niaconis/github-workspace/ember-lifeline/node_modules/.bin/semver as it wasn't installed by /Users/niaconis/github-workspace/ember-lifeline/node_modules/semver
npm WARN gentlyRm not removing /Users/niaconis/github-workspace/ember-lifeline/node_modules/.bin/nopt as it wasn't installed by /Users/niaconis/github-workspace/ember-lifeline/node_modules/nopt
npm WARN gentlyRm not removing /Users/niaconis/github-workspace/ember-lifeline/node_modules/.bin/jsesc as it wasn't installed by /Users/niaconis/github-workspace/ember-lifeline/node_modules/jsesc
npm WARN gentlyRm not removing /Users/niaconis/github-workspace/ember-lifeline/node_modules/fsevents/node_modules/.bin/node-pre-gyp as it wasn't installed by /Users/niaconis/github-workspace/ember-lifeline/node_modules/fsevents/node_modules/node-pre-gyp
npm WARN gentlyRm not removing /Users/niaconis/github-workspace/ember-lifeline/node_modules/ember-cli-dependency-checker/node_modules/.bin/semver as it wasn't installed by /Users/niaconis/github-workspace/ember-lifeline/node_modules/ember-cli-dependency-checker/node_modules/semver
npm notice created a lockfile as package-lock.json. You should commit this file.
+ bower@1.8.2
added 1 package and removed 998 packages in 25.39s
bower ember#lts-2-4         not-cached https://github.com/components/ember.git#lts-2-4
bower ember#lts-2-4            resolve https://github.com/components/ember.git#lts-2-4
bower ember#lts-2-4           download https://github.com/components/ember/archive/lts-2-4.tar.gz
bower ember#lts-2-4            extract archive.tar.gz
bower ember#lts-2-4       invalid-meta for:/var/folders/qz/tgp62wm96jv2qjp780nxzxzw000dvf/T/niaconis/bower/98e89bbbb4878ec0cc7833f0f91276c2-18265-jEodhV/bower.json
bower ember#lts-2-4       invalid-meta The "main" field has to contain only 1 file per filetype; found multiple .js files: ["./ember.debug.js","./ember-template-compiler.js"]
bower ember#lts-2-4           resolved https://github.com/components/ember.git#ea11df0ad7
bower jquery#>= 1.7.0 < 3.0.0   cached https://github.com/jquery/jquery-dist.git#2.2.4
bower jquery#>= 1.7.0 < 3.0.0 validate 2.2.4 against https://github.com/jquery/jquery-dist.git#>= 1.7.0 < 3.0.0
bower ember           extra-resolution Unnecessary resolution: ember#lts-2-4
bower ember#lts-2-4            install ember#ea11df0ad7
bower jquery#>= 1.7.0 < 3.0.0  install jquery#2.2.4

ember#ea11df0ad7 bower_components/ember
└── jquery#2.2.4

jquery#2.2.4 bower_components/jquery

=== Scenario: ember-lts-2.4 ====================================================

Error!
The command ember test exited Error: Cannot find module 'ember-cli' from '/Users/niaconis/github-workspace/ember-lifeline'
undefined
```

**Comment by scalvert** [comment]
I'll test locally and see what I get.

**Comment by mixonic** [comment]
FYI https://github.com/nickiaconis/ember-lifeline/pull/1

**Comment by rwjblue** [comment]
@nickiaconis - That error looks like an issue that I've seen with ember-try + npm@5. Can you ensure we are using npm@4 in CI to confirm that fixes it?

**Comment by nickiaconis** [comment]
@rwjblue Yes, I had `npm@5.3.0` installed. Installing 4.6.1 fixed the issue. Thanks 🙂

**Comment by rwjblue** [comment]
Status?

**Comment by nickiaconis** [comment]
The only unaddressed item is documentation, which is not entirely true as there are JSDocs for the `scheduleTask` method. What else ought to be documented and where?

File: addon/mixins/run.js [hunk 1]
```diff
@@ -118,11 +120,82 @@ export default Mixin.create({
...
+      };
+    } else if (type === 'string' && this[callbackOrName]) {
+      let callback = this[callbackOrName];
+      task = () => {
+        if (!this.isDestroyed) {
+          callback.call(this, ...args);
+        }
+      };
+    } else {
+      throw new Error('You must pass a callback function or method name to `scheduleTask`.');
```
Comment (rwjblue) [inline addon/mixins/run.js]: Should probably be an assertion

File: addon/mixins/run.js [hunk 2]
```diff
@@ -118,11 +120,82 @@ export default Mixin.create({
...
+    let type = typeof callbackOrName;
+    if (type === 'function') {
+      task = () => {
+        if (!this.isDestroyed) {
+          callbackOrName.call(this, ...args);
+        }
+      };
+    } else if (type === 'string' && this[callbackOrName]) {
+      let callback = this[callbackOrName];
+      task = () => {
```
Comment (rwjblue) [inline addon/mixins/run.js]: This seems identical to the `task` created above, the only difference is the actual callback function.  Would the following work instead:

```js
  let callback;
  if (type === 'function') {
    callback = callbackOrName;
  } else if (type === 'string') {
    callback = this[callbackOrName];
  }

  assert('You must pass a callback function or method name to `scheduleTask`.', typeof callback === 'function');

  let task = () => {
    if (!this.isDestroyed) {
      callback.apply(this, args);
    }
  };
```

File: addon/mixins/run.js [hunk 3]
```diff
@@ -118,11 +120,82 @@ export default Mixin.create({
...
+        if (!this.isDestroyed) {
+          callback.call(this, ...args);
```
Comment (rwjblue) [inline addon/mixins/run.js]: Should probably use `.apply` since we are operating on an array anyways

File: addon/mixins/run.js [hunk 4]
```diff
@@ -118,11 +120,82 @@ export default Mixin.create({
...
+   */
+  scheduleTask(queueName, callbackOrName, ...args) {
+    assert(`Called \`scheduleTask\` without a string as the first argument on ${this}.`, typeof queueName === 'string');
+    assert(`Called \`scheduleTask\` on destroyed object: ${this}.`, !this.isDestroyed);
+
+    let task;
+    let type = typeof callbackOrName;
+    if (type === 'function') {
+      task = () => {
+        if (!this.isDestroyed) {
```
Comment (rwjblue) [inline addon/mixins/run.js]: This probably needs to be modified to use `this.isDestroying || this.isDestroyed`.

Comment (scalvert) [reply addon/mixins/run.js]: 👍

File: addon/mixins/run.js (hunk 2 above)
Comment (scalvert) [reply addon/mixins/run.js]: Totally agree with this. This can be simplified.

**Comment by nickiaconis** [comment]
Great feedback @rwjblue, thanks! I've update the code accordingly.

**Comment by nickiaconis** [comment]
@scalvert @rwjblue That's closer to what I had originally, but @mixonic suggested we take it the the route with no-ops and a single scheduled callback.

If we're backpedaling on that, we can still support Ember < 2.8 with [rwjblue's polyfill](https://github.com/rwjblue/ember-lifeline/pull/14#issuecomment-279725876). It could be added at build time only when Ember < 2.8. Thoughts?

**Comment by scalvert** [comment]
I think we should leave it and only add the polyfill if someone asks for it. Look to the future!

**Comment by mixonic** [comment]
I'm not sure what you have as consensus here. I specifically wrote https://github.com/nickiaconis/ember-lifeline/pull/1/commits/f4dd20a7f651bfe64845ce9a398f9c997d2c0caf in that PR to avoid using the runloop's queue for cancellation. The PR laid out some justification as well: https://github.com/nickiaconis/ember-lifeline/pull/1

**Comment by mixonic** [comment]
To clarify, my concerns are mostly regarding implementation, my position about which is largely unchanged since my original review in February https://github.com/rwjblue/ember-lifeline/pull/14#discussion_r101445203. I don't think these concerns should block merging. 👍

**Comment by scalvert** [comment]
Thanks, @mixonic. I only became aware of your feedback late yesterday (because it was against Nick's fork, not this PR). Definitely didn't want to diverge from what we thought was the right path.

I'm totally happy to merge this and iterate on it going forward.

**Comment by scalvert** [comment]
@nickiaconis if you could add docs this PR will be good to go 👍

**Comment by nickiaconis** [comment]
@scalvert Docs added. Let me know if anything needs revising.

**Comment by nickiaconis** [comment]
@scalvert I've rebased the branch since it had fallen behind the upstream master.

File: README.md [hunk 5]
```diff
@@ -144,6 +144,82 @@ export default Component.extend(RunMixin, {
...
+import Ember from 'ember';
+
+const { Component, run } = Ember;
+
+export default Component.extend({
+  init() {
+    this._super();
+    run.schedule('sync', this, () => {
+      // First, check if this object is even valid
+      if (this.isDestroyed) { return; }
```
Comment (scalvert) [inline README.md]: Can we add `this.isDestroying &&...`

**Pull request merged**